From 0322f76c2addcb5d786f0eb1d8a8007ee33e7b1e Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Fri, 29 May 2015 15:12:23 +0300
Subject: [PATCH] drm: i2c: adv7511_core: Fix ADV7511 subchips offsets

Possible i2c main addresses for adv7511 are 0x39 and 0x3d,
add (0x3d - 0x39) offset for subchips to avoid conflict of 2 similar chips
on the same i2c bus.

signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
---
 arch/arm/boot/dts/r8a7790-startnow.dts             | 461 ++++++++++++
 .../board-r8a7790startnow-reference.c              | 831 +++++++++++++++++++++
 drivers/gpu/drm/i2c/adv7511_core.c                 |  11 +-
 3 files changed, 1300 insertions(+), 3 deletions(-)
 create mode 100644 arch/arm/boot/dts/r8a7790-startnow.dts
 create mode 100644 arch/arm/mach-shmobile/board-r8a7790startnow-reference.c

diff --git a/drivers/gpu/drm/i2c/adv7511_core.c b/drivers/gpu/drm/i2c/adv7511_core.c
index a987b5b..a43e760 100644
--- a/drivers/gpu/drm/i2c/adv7511_core.c
+++ b/drivers/gpu/drm/i2c/adv7511_core.c
@@ -930,9 +930,9 @@ static int adv7511_parse_dt(struct device_node *np, struct adv7511_link_config *
 
 	return 0;
 }
-static const int edid_i2c_addr = 0x7e;
-static const int packet_i2c_addr = 0x70;
-static const int cec_i2c_addr = 0x78;
+static int edid_i2c_addr = 0x7e;
+static int packet_i2c_addr = 0x70;
+static int cec_i2c_addr = 0x78;
 
 static int adv7511_probe(struct i2c_client *i2c,
 	const struct i2c_device_id *id)
@@ -1001,6 +1001,13 @@ static int adv7511_probe(struct i2c_client *i2c,
 	if (ret)
 		return ret;
 
+	/* possible i2c main addresses are 0x39 and 0x3d,
+	 * to avoid conflicts add offset for subchips
+	 */
+	edid_i2c_addr += 4;
+	packet_i2c_addr += 4;
+	cec_i2c_addr += 4;
+
 	regmap_write(adv7511->regmap, ADV7511_REG_EDID_I2C_ADDR, edid_i2c_addr);
 	regmap_write(adv7511->regmap, ADV7511_REG_PACKET_I2C_ADDR, packet_i2c_addr);
 	regmap_write(adv7511->regmap, ADV7511_REG_CEC_I2C_ADDR, cec_i2c_addr);
-- 
1.9.1

